pkg/syncer/namespace/namespace_downstream_process.go

```diff
@@ -148,13 +167,57 @@ func (c *DownstreamController) updateDNSConfigMap(ctx context.Context) error {
 		if apierrors.IsNotFound(err) {
 			_, err = c.createConfigMap(ctx, cm)
 			if err != nil {
-				logger.Error(err, "failed to create ConfigMap (retrying)", "name", nsmap.ConfigMapName, "namespace", c.dnsNamespace)
+				logger.Error(err, "failed to create ConfigMap (retrying)", "name", configMapName, "namespace", c.dnsNamespace)
 				return err // retry
 			}
 		}
 
-		logger.Error(err, "failed to update ConfigMap (retrying)", "name", nsmap.ConfigMapName, "namespace", c.dnsNamespace)
+		logger.Error(err, "failed to update ConfigMap (retrying)", "name", configMapName, "namespace", c.dnsNamespace)
 		return err // retry
 	}
 	return nil
 }
+
+// ensureDNSExists checks all DNS objects exist.
+func (c *DownstreamController) ensureDNSExists(ctx context.Context, workspace logicalcluster.Name) error {
+	logger := klog.FromContext(ctx)
+	logger.WithName("dns")
+
+	svcName := shared.GetDNSID(workspace, c.syncTargetUID, c.syncTargetName)
+	logger.Info("ensuring all dns objects exist", "workspace", svcName)
+
+	// TODO: check objects have the expected content (eg. after an upgrade)
+
+	if err := c.ensureObjectExist(ctx, logger, "ServiceAccount", c.getServiceAccount, c.createServiceAccount, svcName); err != nil {
+		return err
+	}
+	if err := c.ensureObjectExist(ctx, logger, "Role", c.getRole, c.createRole, svcName); err != nil {
+		return err
+	}
+	if err := c.ensureObjectExist(ctx, logger, "RoleBinding", c.getRoleBinding, c.createRoleBinding, svcName); err != nil {
+		return err
+	}
+	if err := c.ensureObjectExist(ctx, logger, "Deployment", c.getDeployment, c.createDeployment, svcName); err != nil {
+		return err
+	}
+	if err := c.ensureObjectExist(ctx, logger, "Service", c.getService, c.createService, svcName); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *DownstreamController) ensureObjectExist(ctx context.Context, logger klog.Logger, kind string, get getFn, create createFn, name string) error {
```

Comment:
Not sure that the factorizing of this typical logic into one function + the abstraction of `getFun` and `createFn` really make the code more readable, especially if now the deployment should have a dedicated logic

Reply:
fixed.